### PR TITLE
[BUG FIX]: smoother progress for xet downloads

### LINKED ISSUE #4058: [BUG]: xet downloads barely report progress (bar appears stuck on large files)
## Problem

When downloading files via xet storage, the tqdm progress bar barely updates. For a 4GB file, the bar updates about 4 times total -- `0% → 26% → 63% → 100%`, with gaps representing gigabytes of apparent silence. For large downloads this makes it look like the process has hung.

## Root cause

Two layers:

1. **xet-core**: Network transfer progress was tracked internally but never sent to Python callbacks (hardcoded to zero). Fixed in https://github.com/huggingface/xet-core/pull/791.
2. **huggingface_hub**: `xet_get()` and `download_bucket_files()` use a 1-arg callback that only receives a simple byte count with very coarse granularity. Switching to the 2-arg signature gives access to `total_transfer_bytes_completion_increment` which updates per HTTP chunk. Fixed in https://github.com/huggingface/huggingface_hub/pull/4059.

### Single file

```python
"""Verify smooth single-file download progress."""
import tempfile
from huggingface_hub import hf_hub_download

with tempfile.TemporaryDirectory() as tmp:
    hf_hub_download(
        "Qwen/Qwen3-4B",
        filename="model-00002-of-00003.safetensors",
        cache_dir=tmp,
        force_download=True,
    )
```

### Multi-file

```python
"""Verify smooth multi-file download progress."""
import tempfile
from huggingface_hub import snapshot_download

with tempfile.TemporaryDirectory() as tmp:
    snapshot_download(
        "Qwen/Qwen3-4B",
        allow_patterns=["model-00003-of-00003.safetensors", "config.json", "*.txt"],
        cache_dir=tmp,
        force_download=True,
    )
```

**Before**: tqdm bar jumps 0% -> 26% -> 63% -> 100% with ~4 callbacks for a 4GB file.
**After**: tqdm bar updates smoothly every 250ms.

### Current behavior (bar appears stuck)

Current behavior

### Expected behavior (smooth progress)

Expected behavior

> [!NOTE]
> Both downloads finish in roughly the same time (~80s). The difference is purely in how progress is reported.

#### Current behavior (bar barely moves)

Current snapshot progress

#### Expected behavior (smooth per-file progress)

Expected snapshot progress

## Related

- https://github.com/huggingface/xet-core/pull/791 (xet-core fix)
- https://github.com/huggingface/huggingface_hub/pull/4059 (huggingface_hub fix)
- #4051 / #4043 (earlier PRs that identified the granularity problem)
- #4056 (fixes custom `tqdm_class` handling, orthogonal to this)

### PR DESCRIPTION
**Pull request by tobocop2**
Fixes #4058.

> Depends on https://github.com/huggingface/xet-core/pull/791

## Problem

Xet downloads barely report progress. For a 4GB file the bar updates about 4 times total -- `0% → 26% → 63% → 100%`. The gaps represent gigabytes of apparent silence.

The root cause spans two layers:

1. **xet-core**: Network transfer progress was tracked internally but never sent to Python callbacks. Fixed in https://github.com/huggingface/xet-core/pull/791.
2. **huggingface_hub**: `xet_get()` and `download_bucket_files()` use a 1-arg callback that only receives a simple byte count with very coarse granularity. Switching to the 2-arg signature gives access to `total_transfer_bytes_completion_increment` which updates per HTTP chunk.

## Fix

`make_xet_progress_callback(progress_bar, file_size)` returns a 2-arg callback that reads `total_transfer_bytes_completion_increment` for smooth, real-time progress. Transfer bytes are scaled to file size so the bar tracks 0-100% correctly (transfer size can differ from file size due to compression/dedup).

Wired into both `xet_get()` and `download_bucket_files()`. No public API changes.

### Single file

```python
"""Verify smooth single-file download progress."""
import tempfile
from huggingface_hub import hf_hub_download

with tempfile.TemporaryDirectory() as tmp:
    hf_hub_download(
        "Qwen/Qwen3-4B",
        filename="model-00002-of-00003.safetensors",
        cache_dir=tmp,
        force_download=True,
    )
```

### Multi-file

```python
"""Verify smooth multi-file download progress."""
import tempfile
from huggingface_hub import snapshot_download

with tempfile.TemporaryDirectory() as tmp:
    snapshot_download(
        "Qwen/Qwen3-4B",
        allow_patterns=["model-00003-of-00003.safetensors", "config.json", "*.txt"],
        cache_dir=tmp,
        force_download=True,
    )
```

**Before**: tqdm bar jumps 0% -> 26% -> 63% -> 100% with ~4 callbacks for a 4GB file.
**After**: tqdm bar updates smoothly every 250ms.

## Related

- https://github.com/huggingface/xet-core/pull/791 (xet-core fix, required)
- #4058 (issue)
- #4051 / #4043 (earlier PRs that identified the granularity problem)
- #4056 (fixes custom `tqdm_class` handling, orthogonal to this)

---

> [!NOTE]
> **Medium Risk**
> Changes Xet download progress reporting to a new 2-arg callback signature and adds scaling logic from network transfer bytes to expected file sizes, which could affect progress behavior across single and multi-file downloads if the upstream `hf_xet` progress fields differ from expectations.
> 
> **Overview**
> Improves Xet-backed download progress reporting by introducing `make_xet_progress_callback`, a shared helper that consumes xet-core’s 2-argument progress updates and **scales network transfer progress to the expected file size** (handling dedup/compression and capping at 100%).
> 
> Wires this callback into both `xet_get()` and `HfApi.download_bucket_files()` so tqdm bars update smoothly and correctly for single-file and multi-file/bucket downloads, and adds targeted tests covering signature, scaling, edge cases, and shared-progress-bar behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0cb03be94aa54de3ec56b571edb60243e1182009. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by tobocop2** [comment]
Friendly bump on this, happy to address any feedback if needed. Let me know if there's anything blocking review.

File: src/huggingface_hub/file_download.py [hunk 1]
```diff
@@ -56,8 +57,62 @@
...
+    _last_contributed = 0
+
+    def _callback(
+        total_update: "PyTotalProgressUpdate",
+        item_updates: "list[PyItemProgressUpdate]",
+    ) -> None:
+        nonlocal _cumulative_transfer, _last_contributed
+        increment = total_update.total_transfer_bytes_completion_increment
+        if increment <= 0:
+            return
```
[the remark this answers is not in the document]
Comment (tobocop2) [reply src/huggingface_hub/file_download.py:91]: this is wrong, it's an issue with my PR description, not the code itself. The Cursor bot took my PR description ate face value rather than tracing the blast radius.

**Comment by Wauplin** [comment]
Hi @tobocop2 , thanks for the report and PR! Better reporting is indeed a good addition. I'll wait until https://github.com/huggingface/xet-core/pull/791 gets merged and released before movign forward on this PR. I know the Xet team is preparing a big refacto of how Python should call the underlying Xet Session (see https://github.com/huggingface/xet-core/pull/792) so I don't know if it'll impact your PR or not. Let's see how it settles down.
